### PR TITLE
Docker backend

### PR DESCRIPTION
**Pull request by MisterDA**

### CODE REVIEW

File: lib/docker.mli [hunk 1]
```diff
@@ -1,3 +1,64 @@
...
+(** Get the default entrypoint of Docker container according to the
+    host (Windows is cmd, everywhere else is sh). *)
+
+val setup_command : entp:string list -> cmd:string list -> string * string list
+(** [setup_command ~entp ~cmd] returns the head of [entp], to be
+    give to Docker's [--entrypoint], and the concatenation of the tail
+    of [head] and [cmd] to be given to Docker command, as Docker
+    [--entrypoint] takes only one argument. *)
+
+(** Wrappers for various Docker client commands, exposing file descritors. *)
```
Comment (tmcgilchrist) [inline lib/docker.mli:45]:
```suggestion
(** Wrappers for various Docker client commands, exposing file descriptors. *)
```

File: stress/stress.ml [hunk 2]
```diff
@@ -105,7 +103,13 @@ module Test(Store : S.STORE) = struct
...
       Lwt.return_unit
 
-  module Build = Builder(Store)(Sandbox)(Fetcher)
+  type builder = Builder : (module Obuilder.BUILDER with type t = 'a) * 'a -> builder
+
+  let create_builder store conf =
+    let module Builder = Obuilder.Builder(Store)(Native_sandbox)(Obuilder.Docker_extract) in
+    Native_sandbox.create ~state_dir:(Store.state_dir store / "sandbox") conf >|= fun sandbox ->
+    let builder = Builder.v ~store ~sandbox in
+    Builder ((module Builder), builder)
```
Comment (tmcgilchrist) [inline stress/stress.ml:112]: It would be useful to have a stress test for the obuilder on Docker setup.

File: lib/tar_transfer.mli [hunk 3]
```diff
@@ -21,3 +21,27 @@ val send_file :
...
+  from_tar:Lwt_unix.file_descr ->
+  src_manifest:Manifest.t ->
+  dst:string ->
+  user:Obuilder_spec.user ->
+  to_untar:Lwt_unix.file_descr ->
+  unit Lwt.t
+(** [transform_files ~src_dir ~from_tar ~src_manifest ~dst ~user ~to_untar]
+    renames the _unique_ file found in [from_tar], a tar archive streamed in
+    input, to [dst], and writes the resulting tar-format stream to
+    [to_untar]. All files are listed as being owned by [user]. *)
```
Comment (tmcgilchrist) [inline lib/tar_transfer.mli:47]: Add tests for this?

File: lib/s.ml [hunk 4]
```diff
@@ -128,3 +128,106 @@ module type FETCHER = sig
       @param log Used for outputting the progress of the fetch
       @param rootfs The directory in which to extract the base image *)
 end
+
+(** Wrappers for various Docker client commands.  *)
+module type DOCKER_CMD = sig
```
Comment (tmcgilchrist) [inline lib/s.ml:133]: Write documentation comments for this module.

File: lib/docker.ml [hunk 5]
```diff
@@ -1,37 +1,344 @@
...
+
+let exec_result' ?stdin ?stdout ?stderr ?is_success argv =
+  let cmd = "docker" :: argv in
+  let pp f = Os.pp_cmd f ("", cmd) in
+  let stdin = Option.value ~default:`Dev_null stdin in
+  let stdout = Option.value ~default:(`FD_move_safely Os.stdout) stdout in
+  let stderr = Option.value ~default:(`FD_move_safely Os.stderr) stderr in
+  Os.exec_result ~stdin ~stdout ~stderr ?is_success ~pp cmd
+
+module Cmd = struct
```
Comment (tmcgilchrist) [inline lib/docker.ml:70]: Good to have integration tests for these commands. Run against docker linux, macos and windows.

File: lib/docker.ml [hunk 6]
```diff
@@ -1,37 +1,344 @@
...
+    with_stderr_log ~log (fun ~stderr ->
+        Cmd.run' ?stdin ?stdout ~stderr ?is_success ?name ?rm docker_argv image argv)
+
+  let run_result ?stdin ~log ?name ?rm docker_argv image argv =
+    with_log ~log (fun ~stdout ~stderr ->
+        Cmd.run_result ?stdin ~stdout ~stderr ?name ?rm docker_argv image argv)
+
+  let run_result' ?stdin ?stdout ~log ?name ?rm docker_argv image argv =
+    with_stderr_log ~log (fun ~stderr ->
+        Cmd.run_result' ?stdin ?stdout ~stderr ?name ?rm docker_argv image argv)
```
Comment (tmcgilchrist) [inline lib/docker.ml:286]: Why have these almost duplicate functions?

File: lib/build.mli [hunk 7]
```diff
@@ -27,3 +27,9 @@ module Make (Store : S.STORE) (Sandbox : S.SANDBOX) (_ : S.FETCHER) : sig
 
   val v : store:Store.t -> sandbox:Sandbox.t -> t
 end
+
+module Make_Docker (Store : S.STORE) : sig
+  include S.BUILDER with type context := Context.t
+
+  val v : store:Store.t -> sandbox:Docker_sandbox.t -> t
+end
```
Comment (tmcgilchrist) [inline lib/build.mli:35]: Question here, why is this necessary and not just have a Docker version of S.STORE and S.SANDBOX?

**Comment by MisterDA** [comment]
_Alea jacta est_
